pyro/infer/mcmc/nuts.py

```diff
@@ -115,19 +116,27 @@ def __init__(self,
         # Here, as suggested in [1], we set dE_max = 1000.
         self._max_sliced_energy = 1000
 
-    def _is_turning(self, z_left, r_left, z_right, r_right):
-        diff_left = 0
-        diff_right = 0
-        for name in self._r_shapes:
-            dz = z_right[name] - z_left[name]
-            diff_left += (dz * r_left[name]).sum()
-            diff_right += (dz * r_right[name]).sum()
-        return diff_left < 0 or diff_right < 0
+        # Set a flag to decide if we want to eliminate the initial point from the candidates to
+        # choose uniformly along the trajectory. In [1], this flag is True, but in Stan, they set
+        # it to False (implicitly).
+        self._eliminate_starting_point = True
+
+    def _is_turning(self, r_left, r_right, r_sum):
+        # We follow the strategy in Section A.4.2 of [2] for this implementation.
```

Comment:
Would also suggest including this [reference](https://arxiv.org/pdf/1304.1920.pdf) which derives the termination criterion explicitly.

Reply:
In [2], the author also derives the criterion too. I find it easier to understand than differential geometry style (though it is a nice language).

Reply:
Oh..its the same derivation. I was looking at 4.2 rather than the appendix.